.github/workflows/deploy.yml

```
@@ -0,0 +1,52 @@
+name: Deploy
+
+on:
+  push:
+    branches:
+      - master
+      - source
+  pull_request:
+    branches:
+      - master
+      - source
+
+jobs:
+  deploy:
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout code
+      uses: actions/checkout@v2
+    - name: Setup Ruby
+      uses: actions/setup-ruby@v1
+    - name: Enable bundler cache
+      uses: actions/cache@v2
+      with:
+        path: vendor/bundle
+        key: ${{ runner.os }}-gems-${{ hashFiles('**/Gemfile.lock') }}
+        restore-keys: |
+          ${{ runner.os }}-gems-
+    - name: Install deps
+      run: |
+        gem install bundler
+        bundle config path vendor/bundle
+        bundle install --jobs 4 --retry 3
+    - name: Setup deploy options
+      id: setup
+      run: |
+        git config --global user.name "GitHub Action"
+        git config --global user.email "41898282+github-actions[bot]@users.noreply.github.com"
+        if [[ ${GITHUB_REF} = refs/pull/*/merge ]]; then # pull request
+          echo "::set-output name=SRC_BRANCH::${GITHUB_BASE_REF}"
+          echo "::set-output name=DRY_RUN::--dry-run"
+        elif [[ ${GITHUB_REF} = refs/heads/* ]]; then # branch, e.g. master, source etc
+          echo "::set-output name=SRC_BRANCH::${GITHUB_REF#refs/heads/}"
+        fi
```

Could you explain the logic of this step:

1) Do I understand correctly that on PR, it will set `SRC_BRANCH` to `${GITHUB_BASE_REF}` (which must be the PR target) and also set the `--dry-run` option on. Why is dry-run necessary? What is the intended behavior on PR?
2) On push, it sets `SRC_BRANCH` to the branch name the push was made to, correct?

---

Re:  1. I was actually mistaken here, `GITHUB_BASE_REF` is the target branch, so we need to use `GITHUB_HEAD_REF` here (see https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions). The `--dry-run` is needed, so that a PR from an internal branch doesn't get deployed. PR from a fork won't have write access to the repo, so there it won't be an issue.

Re: 2. Yes, the action is only triggered when pushed to `master` or `source` and then `GITHUB_REF` will be something like `refs/heads/master`. It is worth mentioning that, in the user/org case, a push to the `source` branch will trigger the deploy action onto master, but the then automatic commit to the `master` branch won't trigger another Github action execution. GitHub doesn't allow a push triggered by an action to trigger another action.

---

Thanks for explaining!

Re: 1. In the dry-run regime, I guess base vs. head branch doesn't matter. Just to double-check the logic, on internal PR merge, the action will get triggered twice: (1)  due to PR (which will be a dry-run) and (2) due to the resulting commit from the PR in the base branch (which will trigger the actual deploy), right?

---

Re 1. I still think it needs to be `GITHUB_HEAD_REF`, let's assume you have PR from `new_feature` to `master`. The action is supposed to test the commits on `new_feature` (`GITHUB_HEAD_REF`) and not `master` (`GITHUB_HEAD_REF`). And this PR won't trigger twice as a push to `new_feature` doesn't trigger an action, only an open PR will trigger an action (see the selection of branches in the top of the file `on: push: branches:`)

---

Plus I will submit a bunch of unrelated other changes as PR afterwards to test this ;-)

---

> Re 1. I still think it needs to be `GITHUB_HEAD_REF`, let's assume you have PR from `new_feature` to `master`. The action is supposed to test the commits on `new_feature` (`GITHUB_HEAD_REF`) and not `master` (`GITHUB_HEAD_REF`). And this PR won't trigger twice as a push to `new_feature` doesn't trigger an action, only an open PR will trigger an action (see the selection of branches in the top of the file `on: push: branches:`)

Got it, so the following adjustment should make it work as expected?
```suggestion
        if [[ ${GITHUB_REF} = refs/pull/*/merge ]]; then # pull request
          echo "::set-output name=SRC_BRANCH::${GITHUB_HEAD_REF}"
          echo "::set-output name=DRY_RUN::--dry-run"
        elif [[ ${GITHUB_REF} = refs/heads/* ]]; then # branch, e.g. master, source etc
          echo "::set-output name=SRC_BRANCH::${GITHUB_REF#refs/heads/}"
        fi
```

> Plus I will submit a bunch of unrelated other changes as PR afterwards to test this ;-)

Awesome, thanks! :)